File: 2024_12_06_Python_Vs_Java_Challenges/Challenge_2/numbers.py
class Numbers:
    def __init__(self, numbers: list[int] = None):
        if numbers is None:
            self._numbers = []
        else:
            self._numbers = numbers
        self._numbers.sort()

    def add_number(self, number: int):
        self._numbers.append(number)
        self._numbers.sort()

    def remove(self, number: int):
        self._numbers.remove(number)

    def get_max(self):
        return self._numbers[-1]

    def get_min(self):
        return self._numbers[0]

    def remove_duplicates(self):
        self._numbers = list(set(self._numbers))
        self._numbers.sort()

    def __str__(self):
        return " ".join(str(number) for number in self._numbers).strip()

if __name__ == "__main__":
    my_numbers = Numbers()
    my_numbers.add_number(10)
    my_numbers.add_number(1)
    my_numbers.add_number(5)
    my_numbers.add_number(10)
    my_numbers.add_number(25)
    my_numbers.add_number(10)
    my_numbers.add_number(-25)
    print(my_numbers) # -25 1 5 10 10 10 25
    print(my_numbers.get_max()) # 25
    print(my_numbers.get_min()) # -25
    my_numbers.remove_duplicates()
    print(my_numbers) # -25 1 5 10 25

File: 2024_12_06_Python_Vs_Java_Challenges/Challenge_2/Numbers.java
// package com.gradescope.numbers;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;

public class Numbers {
    private ArrayList<Integer> numbers = new ArrayList<Integer>();

    public Numbers() {

    }

    public void addNumber(int number) {
        numbers.add(number);
        numbers.sort(Integer::compareTo);
    }

    public void remove(int number) {
        numbers.sort(Integer::compareTo);
        numbers.remove(number);
    }

    public int getMax() {
        numbers.sort(Integer::compareTo);
        return numbers.get(numbers.size() - 1);
    }

    public int getMin() {
        numbers.sort(Integer::compareTo);
        return numbers.get(0);
    }

    public void removeDuplicates() {
        Set<Integer> numberSet = new HashSet<>(numbers); // Use a HashSet to remove duplicates
        numbers = new ArrayList<>(numberSet); // Convert back to a List
        numbers.sort(Integer::compareTo);
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        for(int number: numbers) {
            sb.append(number + " ");
        }
        return sb.toString().strip();
    }

    public static void main(String[] args) {
        Numbers myNumbers = new Numbers();
        myNumbers.addNumber(10);
        myNumbers.addNumber(1);
        myNumbers.addNumber(5);
        myNumbers.addNumber(10);
        myNumbers.addNumber(25);
        myNumbers.addNumber(10);
        myNumbers.addNumber(-25);
        System.out.println(myNumbers); // -25 1 5 10 10 10 25
        myNumbers.removeDuplicates();
        System.out.println(myNumbers); // -25 1 5 10 25
    }
}